lib/history/sqlite/sqlite.go

```diff
@@ -30,115 +29,142 @@ import (
+func (t *Timeline) initSQLite(ctx context.Context) error {
+	database, err := sqlx.ConnectContext(ctx, "sqlite3", t.config.DBPath)
 	if err != nil {
-		return nil, trace.Wrap(err)
+		return trace.Wrap(err)
 	}
 
 	if _, err := database.ExecContext(ctx, createTableEvents); err != nil {
-		return nil, trace.Wrap(err)
+		return trace.Wrap(err)
 	}
 
-	return database, nil
+	t.database = database
+	return nil
+}
+
+// initPrevStatus initializes the previously record status in case satellite
+// is restarted.
+func (t *Timeline) initPrevStatus(ctx context.Context) error {
+	// TODO
+	return trace.NotImplemented("not implemented")
 }
 
-// getSize returns the current number of rows stored in the provided database.
-func getSize(ctx context.Context, database *sqlx.DB) (size int, err error) {
-	row := database.QueryRowContext(ctx, "SELECT COUNT(*) FROM events")
-	if err := row.Scan(&size); err != nil {
-		return -1, trace.Wrap(err)
+// eventEvictionLoop periodically evicts old events to free up storage.
+func (t *Timeline) eventEvictionLoop() {
+	ticker := t.config.Clock.NewTicker(evictionFrequency)
+	for {
+		<-ticker.Chan()
+
+		// All events before this time will be deleted.
+		retentionCutOff := t.config.Clock.Now().Add(-(t.config.RetentionDuration))
+
+		ctx, cancel := context.WithTimeout(context.Background(), evictionTimeout)
+		if err := t.evictEvents(ctx, retentionCutOff); err != nil {
+			cancel()
+			log.WithError(err).Warnf("Error evicting expired events.")
+		}
+		cancel()
 	}
-	return size, nil
 }
 
 // RecordStatus records the differences between the previously stored status and
 // the provided status.
-func (t *Timeline) RecordStatus(ctx context.Context, status *pb.SystemStatus) (err error) {
-	t.mu.Lock()
+func (t *Timeline) RecordStatus(ctx context.Context, status *pb.NodeStatus) (err error) {
+	t.statusMutex.Lock()
 
-	events := history.DiffCluster(t.Config.Clock, t.lastStatus, status)
+	events := history.DiffNode(t.config.Clock, t.lastStatus, status)
 	if len(events) == 0 {
-		t.mu.Unlock()
+		t.statusMutex.Unlock()
 		return nil
 	}
-
-	t.size = t.size + len(events)
-	if t.size > t.Config.Capacity {
-		t.size = t.Config.Capacity
-	}
 	t.lastStatus = status
 
-	t.mu.Unlock()
+	t.statusMutex.Unlock()
 
 	if err = t.insertEvents(ctx, events); err != nil {
-		return trace.Wrap(err, "failed to insert events")
+		return trace.Wrap(err)
 	}
 
-	if t.size == t.Config.Capacity {
-		if err = t.evictEvents(ctx); err != nil {
-			return trace.Wrap(err, "failed to evict old events")
+	return nil
+}
+
+// RecordTimeline merges the provided events into the current timeline.
+// Duplicate events will be ignored.
+func (t *Timeline) RecordTimeline(ctx context.Context, events []*pb.TimelineEvent) (err error) {
+	if len(events) == 0 {
+		return nil
+	}
+
+	// All events before this time will be deleted.
+	retentionCutOff := t.config.Clock.Now().Add(-(t.config.RetentionDuration))
```

Comment:
Nit: I notice this incantation is used here for the second time, I would consider moving it out into a single-line helper method, like `func getRetentionCutoff() time.Time` or something.